File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
private:
    vector<string> result;
    
public:
    void backtrack(int n, int openN, int closeN, string curr){
        if(n == openN && closeN == n){
            result.push_back(curr);
            return;
        }
        
        if(openN < n){
            backtrack(n, openN + 1, closeN, curr + "(");
        }
        
        if(closeN < openN){
            backtrack(n, openN, closeN + 1, curr + ")");
        }
    }
    
    vector<string> generateParenthesis(int n) {
        result.clear();
        backtrack(n, 0, 0, "");
        return result;
    }
};

File: 0022-generate-parentheses/0022-generate-parentheses.py
result = "";


class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        result = []
        def backtrack(openN, closeN, curr):
            if openN == n and closeN == n:
                result.append(curr);
    
            if openN < n:
                backtrack(openN + 1, closeN, curr + '(');
    
            if openN > closeN:
                backtrack(openN, closeN + 1, curr + ')');
        
        backtrack(0, 0, "")
        return result
        